src/proxy/http/fallback.rs

```diff
@@ -0,0 +1,334 @@
+    Layer {
+        primary,
+        fallback,
+        _p: PhantomData,
+    }
+}
+
+impl<P, F, A, M> svc::Layer<M> for Layer<P, F, A>
+where
+    P: svc::Layer<M> + Clone,
+    F: svc::Layer<M> + Clone,
+    M: Clone,
+{
+    type Service = MakeFallback<P::Service, F::Service, A>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        MakeFallback {
+            primary: self.primary.clone().service(inner.clone()),
+            fallback: self.fallback.clone().service(inner),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A, T> svc::Service<T> for MakeFallback<P, F, A>
+where
+    P: svc::Service<T>,
+    P::Error: Into<proxy::Error>,
+    F: svc::Service<T>,
+    F::Error: Into<proxy::Error>,
+    T: Clone,
+{
+    type Response = Service<P::Response, F::Response, A>;
+    type Error = proxy::Error;
+    type Future = MakeFuture<P::Future, F::Future, A>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        let primary_ready = self.primary.poll_ready().map_err(Into::into);
+        let fallback_ready = self.fallback.poll_ready().map_err(Into::into);
+        try_ready!(fallback_ready);
+        primary_ready
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        MakeFuture {
+            primary: Making::NotReady(self.primary.call(target.clone())),
+            fallback: Making::NotReady(self.fallback.call(target.clone())),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A> Clone for MakeFallback<P, F, A>
+where
+    P: Clone,
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            primary: self.primary.clone(),
+            fallback: self.fallback.clone(),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A> Future for MakeFuture<P, F, A>
+where
+    P: Future,
+    P::Error: Into<proxy::Error>,
+    F: Future,
+    F::Error: Into<proxy::Error>,
+{
+    type Item = Service<P::Item, F::Item, A>;
+    type Error = proxy::Error;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Are both the primary and fallback futures finished?
+        try_ready!(self.primary.poll().map_err(Into::into));
+        try_ready!(self.fallback.poll().map_err(Into::into));
+
+        Ok(Async::Ready(Service {
+            primary: self.primary.take(),
+            fallback: self.fallback.take(),
+            _p: PhantomData,
+        }))
+    }
+}
+
+impl<T: Future> Making<T> {
+    fn poll(&mut self) -> Poll<(), T::Error> {
+        *self = match self {
+            Making::NotReady(ref mut fut) => {
+                let res = try_ready!(fut.poll());
+                Making::Ready(res)
+            }
+            Making::Ready(_) => return Ok(Async::Ready(())),
+            Making::Done => panic!("polled after ready"),
+        };
+        Ok(Async::Ready(()))
+    }
+
+    fn take(&mut self) -> T::Item {
+        match mem::replace(self, Making::Done) {
+            Making::Ready(a) => a,
+            _ => panic!(),
```

Comment:
A comment here could be useful. Something like `tried to take service after ready`?